src/mesh/p4est_mesh.jl

```diff
@@ -0,0 +1,273 @@
+                   initial_refinement_level=initial_refinement_level,
+                   periodicity=periodicity, mapping_as_string=mapping_as_string)
+end
+
+
+"""
+    P4estMesh(trees_per_dimension, coordinates_min, coordinates_max;
+              polydeg, initial_refinement_level=0, periodicity=true)
+
+Create a P4estMesh that represents an uncurved structured mesh with a rectangular domain.
+
+# Arguments
+- `trees_per_dimension::NTuple{NDIMS, Int}`: the number of trees in each dimension.
+- `coordinates_min::NTuple{NDIMS, RealT}`: coordinate of the corner in the negative direction of each dimension.
+- `coordinates_max::NTuple{NDIMS, RealT}`: coordinate of the corner in the positive direction of each dimension.
+- `polydeg::Integer`: polynomial degree to used to store the geometry of the mesh.
+                      The mapping will be approximated by an interpolation polynomial
+                      of the specified degree for each tree.
+- `initial_refinement_level::Integer`: refine the mesh uniformly to this level before the simulation starts.
+- `periodicity`: either a `Bool` deciding if all of the boundaries are periodic or an `NTuple{NDIMS, Bool}` deciding for
+                 each dimension if the boundaries in this dimension are periodic.
+"""
+function P4estMesh(trees_per_dimension, coordinates_min, coordinates_max;
+                   polydeg, initial_refinement_level=0, periodicity=true)
+  RealT = promote_type(eltype(coordinates_min), eltype(coordinates_max))
+  mapping = coordinates2mapping(coordinates_min, coordinates_max)
+
+  return P4estMesh(trees_per_dimension, mapping, polydeg=polydeg; RealT=RealT,
+                   initial_refinement_level=initial_refinement_level, periodicity=periodicity)
+end
+
+
+# Create a new p4est_connectivity, which represents a structured rectangle with periodic boundaries.
+# Similar to p4est_connectivity_new_brick, but doesn't use Morton ordering
+# TODO non-periodic
+function connectivity_structured_periodic(cells_x, cells_y)
+  linear_indices = LinearIndices((cells_x, cells_y))
+
+  # Vertices represent the coordinates of the forest. This is used by p4est
+  # to write VTK files.
+  # Trixi doesn't use p4est's coordinates, so the vertices can be empty.
+  num_vertices = 0
+  num_trees = cells_x * cells_y
+  # No corner connectivity is needed
+  num_corners = 0
+  vertices = C_NULL
+  tree_to_vertex = C_NULL
+
+  # Periodic boundaries
+  tree_to_tree = Matrix{Int32}(undef, 4, num_trees)
+  for cell_y in 1:cells_y, cell_x in 1:cells_x
+    tree = linear_indices[cell_x, cell_y]
+    tree_to_tree[1, tree] = linear_indices[mod(cell_x - 2, cells_x) + 1, cell_y] - 1
+    tree_to_tree[2, tree] = linear_indices[mod(cell_x, cells_x) + 1, cell_y] - 1
+    tree_to_tree[3, tree] = linear_indices[cell_x, mod(cell_y - 2, cells_y) + 1] - 1
+    tree_to_tree[4, tree] = linear_indices[cell_x, mod(cell_y, cells_y) + 1] - 1
+  end
+
+  tree_to_face = Matrix{Int8}(undef, 4, num_trees)
```

Comment:
```suggestion
  tree_to_face = Matrix{Cchar}(undef, 4, num_trees)
```
Is that what p4est uses? I would like to keep the C interface as mimetic as possible

Reply:
Yes, p4est uses `int8_t`.